### PR TITLE
Move remains of HackyAI to BotModules

### LINKED ISSUE #15968: AI doesn't account for Transforms.Offset when trying to deploy MCVs
This is the reason it often chooses unsuitable locations, despite otherwise properly using `World.CanPlaceBuilding`.

Will be fixed in #15922.

### PR DESCRIPTION
**Pull request by reaperrr**
This PR splits the remaining parts of HackyAI to a `SquadManagerBotModule`, `UnitBuilderBotModule` and `McvManagerBotModule` and dissolves the AI namespace.

Note: The problem that AIs have trouble attacking when shroud/fog are enabled and the target is out of sight isn't new and not caused by this PR.

~Depends on #15921.~

Fixes #15968.

### CODE REVIEW

File: OpenRA.Mods.Common/OpenRA.Mods.Common.csproj [hunk 1]
```diff
@@ -120,22 +120,24 @@
...
-    <Compile Include="AI\Squad.cs" />
-    <Compile Include="AI\StateMachine.cs" />
-    <Compile Include="AI\States\AirStates.cs" />
-    <Compile Include="AI\States\GroundStates.cs" />
-    <Compile Include="AI\States\NavyStates.cs" />
-    <Compile Include="AI\States\ProtectionStates.cs" />
-    <Compile Include="AI\States\StateBase.cs" />
+    <Compile Include="Traits\BotModules\SquadManagerBotModule.cs" />
+    <Compile Include="Traits\BotModules\UnitBuilderBotModule.cs" />
+    <Compile Include="Traits\BotModules\BotModuleLogic\Squad.cs" />
```
Comment (pchote) [inline OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]: IMO these would be better in a `Squads` subdir rather than `BotModuleLogic`.

File: OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/AttackOrFleeFuzzy.cs [hunk 2]
```diff
@@ -13,11 +13,10 @@
 using System.Collections.Generic;
 using System.Linq;
 using AI.Fuzzy.Library;
-using OpenRA.Mods.Common.Traits;
 using OpenRA.Mods.Common.Warheads;
 using OpenRA.Traits;
 
-namespace OpenRA.Mods.Common.AI
+namespace OpenRA.Mods.Common.Traits
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/BotModuleLogic/AttackOrFleeFuzzy.cs]: And then it would be good to put these in a `OpenRA.Mods.Common.Traits.BotModules.Squads` namespace.

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Controls AI unit production.")]
+	public class UnitBuilderBotModuleInfo : ConditionalTraitInfo
+	{
+		public readonly HashSet<string> McvTypes = new HashSet<string>();
+		public readonly HashSet<string> ConstructionYardTypes = new HashSet<string>();
+		public readonly HashSet<string> McvFactoryTypes = new HashSet<string>();
+		public readonly HashSet<string> RefineryTypes = new HashSet<string>();
+		public readonly HashSet<string> PowerTypes = new HashSet<string>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs]: I'm hoping that most of these can go away from here after implementing the interface suggestions below:

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+		}
+
+		void ProductionUnits(IBot bot)
+		{
+			// Stop building until economy is restored
+			if (!HasAdequateRefineryCount)
+				return;
+
+			// No construction yards - Build a new MCV
+			if (Info.McvTypes.Any() && HasAdequateConstructionYardCount &&
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs]: Can we please generalize this by defining an `IBotRequestUnitProduction` interface implemented by this module which adds the unit type to a queue, and then builds those here instead of hardcoding MCV?

We can then split out a separate `McvManagerBotModule` to manage MCV production and deployment logic.

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+			ticks++;
+
+			if (ticks % FeedbackTime == 0)
+				ProductionUnits(bot);
+		}
+
+		void ProductionUnits(IBot bot)
+		{
+			// Stop building until economy is restored
+			if (!HasAdequateRefineryCount)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs]: Can we please generalize this by introducing an `IBotRequestPauseUnitProduction` interface implemented by `BaseBuilderBotModule` and keep the building-count/prereq logic there?

**Comment by reaperrr** [comment]
Rebased and updated.

**Comment by reaperrr** [comment]
Updated as discussed on IRC.

File: OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs [hunk 6]
```diff
@@ -240,5 +244,26 @@ bool IsRallyPointValid(CPos x, BuildingInfo info)
...
+					AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) == 0;
+			}
+		}
+
+		int MinimumRefineryCount
+		{
+			get
+			{
+				// Unless we have no barracks (higher priority), require a 2nd refinery.
+				// TODO: Possibly unhardcode this, at least the targeted minimum of 2 (the fallback can probably stay at 1).
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/BaseBuilderBotModule.cs]: General thought: IMO it would be better to define a list of build orders for main and expansion bases instead of random order + fraction / limit / priority bodges.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+		// MCVs that the bot already knows about. Any MCV not on this list needs to be given an order.
+		List<Actor> activeMCVs = new List<Actor>();
+
+		public McvManagerBotModule(Actor self, McvManagerBotModuleInfo info)
+			: base(info)
+		{
+			world = self.World;
+			player = self.Owner;
+
+			if (world.Type == WorldType.Editor)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: What's the purpose of this, and why don't we need it in the other module types?

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+					if (unitBuilder != null)
+					{
+						var mcvInfo = AIUtils.GetInfoByCommonName(Info.McvTypes, player);
+						unitBuilder.RequestUnitProduction(bot, mcvInfo.Name);
+					}
+				}
+			}
+		}
+
+		bool ShouldBuildMCV
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: Please make this a method instead of a property.  Generally speaking, get-properties are used in cases where the result can be thought of like a field: so no side-effects, and low performance overheads.  `CountActorByCommonName` can be quite heavy, so this breaks the second expectation.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+		{
+			get
+			{
+				// Only build MCV if we don't already have one in the field.
+				var allowedToBuildMCV = AIUtils.CountActorByCommonName(Info.McvTypes, player) == 0;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: Can we skip `hasAdequateContructionYards` and return early if `!allowedToBuildMCV`?

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+				// Require at least one construction yard, unless we have no vehicles factory (can't build it).
+				var hasAdequateContructionYards = AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) > 0 ||
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: The just return this directly (negated) instead of assigning to a variable.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+
+			foreach (var mcv in activeMCVs)
+				DeployMcv(bot, mcv, chooseLocation);
+		}
+
+		// Find any MCV and deploy them at a sensible location.
+		void DeployMcv(IBot bot, Actor mcv, bool move)
+		{
+			// Don't try to move and deploy an undeployable actor
+			var transformsInfo = mcv.Info.TraitInfoOrDefault<TransformsInfo>();
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: Can we include this in the `newMCVs` filtering instead? This will prevent a bunch of useless references junking up `activeMCVs` and eating perf in mods that require this check.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+				if (desiredLocation == null)
+					return;
+
+				bot.QueueOrder(new Order("Move", mcv, Target.FromCell(world, desiredLocation.Value), true));
+			}
+
+			// If the MCV has to move first, we can't be sure it reaches the destination alive, so we only
+			// update base and defense center if the MCV is deployed immediately (i.e. at game start).
+			// TODO: This could be adressed via INotifyTransform.
+			Sync.RunUnsynced(Game.Settings.Debug.SyncCheckBotModuleCode, world, () =>
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: `DeployMcv` is always called via `IBotTick.BotTick`, so all this code is already running in a `RunUnsynced` context.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 13]
```diff
@@ -0,0 +1,357 @@
...
+		int attackForceTicks;
+		int minAttackForceDelayTicks;
+
+		public SquadManagerBotModule(Actor self, SquadManagerBotModuleInfo info)
+			: base(info)
+		{
+			World = self.World;
+			Player = self.Owner;
+
+			if (World.Type == WorldType.Editor)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Likewise here.

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 14]
```diff
@@ -0,0 +1,357 @@
...
+		}
+
+		void CleanSquads()
+		{
+			Squads.RemoveAll(s => !s.IsValid);
+			foreach (var s in Squads)
+				s.Units.RemoveAll(unitCannotBeOrdered);
+		}
+
+		// Use of this function requires that there is one squad of this type. Hence it is a piece of shit
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Style nit: Can we reword this to `// HACK: Use of this function requires that there is one squad of this type.`?

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 15]
```diff
@@ -0,0 +1,357 @@
...
+			if (--minAttackForceDelayTicks <= 0)
+			{
+				minAttackForceDelayTicks = Info.MinimumAttackForceDelay;
+				CreateAttackForce(bot);
+			}
+		}
+
+		void FindNewUnits(IBot bot)
+		{
+			var newUnits = World.ActorsHavingTrait<IPositionable>()
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: General thought for future improvements: This will pick up all trees/rocks/civ buildings on the map. If we can find a good excuse to introduce some kind of `ControllableByBot` trait with actor-specific data then we could filter on that here instead.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: But chances are that this trait would be introduced as part of a proper replacement for `SquadManagerBotModule`, so ¯\\_(ツ)_/¯

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+				.Where(a => a.Owner == Player && !activeUnits.Contains(a));
+
+			foreach (var a in newUnits)
+			{
+				if (Info.ExcludeFromSquadsTypes.Contains(a.Info.Name))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Can we include this check in the filtering just above?

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs [hunk 17]
```diff
@@ -0,0 +1,357 @@
...
+		}
+
+		void TryToRushAttack(IBot bot)
+		{
+			var allEnemyBaseBuilder = AIUtils.FindEnemiesByCommonName(Info.ConstructionYardTypes, Player);
+			var ownUnits = activeUnits
+				.Where(unit => unit.IsIdle && unit.Info.HasTraitInfo<AttackBaseInfo>()
+					&& !unit.Info.HasTraitInfo<AircraftInfo>() && !unit.Info.HasTraitInfo<HarvesterInfo>()).ToList();
+
+			if (!allEnemyBaseBuilder.Any() || (ownUnits.Count < Info.SquadSize))
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: Unnecessary `()` around the second check.

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+			if (ticks % FeedbackTime == 0)
+				ProductionUnits(bot);
+		}
+
+		void IBotRequestUnitProduction.RequestUnitProduction(IBot bot, string requestedActor)
+		{
+			queuedBuildRequests.Add(requestedActor);
+		}
+
+		void ProductionUnits(IBot bot)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs]: IMO it would be clearer if this was inlined in the one place it is called above.

File: OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs [hunk 19]
```diff
@@ -0,0 +1,218 @@
...
+		{
+			foreach (var r in queuedBuildRequests)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/UnitBuilderBotModule.cs]: Prefer using `queuedBuildRequests.FirstOrDefault()` over a loop that can run at most once.

File: OpenRA.Mods.Common/TraitsInterfaces.cs [hunk 20]
```diff
@@ -463,6 +462,24 @@ public interface IBotPositionsUpdated
 		void UpdatedDefenseCenter(CPos newLocation);
 	}
 
+	[RequireExplicitImplementation]
+	public interface IBotIdleBaseUnits
+	{
+		void UpdatedUnitsHangingAroundBaseCount(int newCount);
```
Comment (pchote) [inline OpenRA.Mods.Common/TraitsInterfaces.cs]: Reporting the number of idle units without having a way to see what they are doesn't make sense for any reasonable future implementations.

How about we change this to `IBotNotifyIdleBaseUnits` and passing the actual list instead?

**Review by pchote — COMMENTED** [review]
A few more thoughts on the code:

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs (hunk 7 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: I don't remember, tbh. Presumably a left-over from outdated module plumbing that got copy-pasted around. I'll try removing it and see what happens (probably nothing).

File: OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs (hunk 15 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/BotModules/SquadManagerBotModule.cs]: I think it's unrealistic for mods to have MCVs that don't implement `IMove` somewhere, so I'll go for that instead.
Modders that really don't use any trait implementing `IMove` for MCVs should be able to write/customize their own McvManager module, too.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs (hunk 7 above)
Comment (reaperrr) [reply OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: Removed it, didn't notice any regressions in the editor.

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+			// Build MCV if we have no construction yard, unless we have no factory (can't build it).
+			return AIUtils.CountBuildingByCommonName(Info.ConstructionYardTypes, player) == 0 &&
+				AIUtils.CountBuildingByCommonName(Info.McvFactoryTypes, player) > 0;
+		}
+
+		void DeployMcvs(IBot bot, bool chooseLocation)
+		{
+			activeMCVs.RemoveAll(unitCannotBeOrdered);
+
+			var newMCVs = world.ActorsHavingTrait<IPositionable>()
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: Can dramatically reduce the set of actors to check by querying `ActorsHavingTrait<Transforms>` instead!

File: OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+				.Where(a => a.Owner == player &&
+					a.IsIdle &&
+					Info.McvTypes.Contains(a.Info.Name) &&
+					a.Info.HasTraitInfo<TransformsInfo>() &&
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/BotModules/McvManagerBotModule.cs]: This line can then be dropped

**Review by pchote — COMMENTED** [review]
One last small code change, then I think the code details are probably reasonable.
This doesn't rule out fundamental logical issues - will need testing to check for these.

**Comment by reaperrr** [comment]
Updated.

**Comment by pchote** [comment]
Definitely not in this PR

**Comment by reaperrr** [comment]
One more tiny update: I made the minimum number of construction yards that a bot aims for configurable (instead of hardcoding 1), which also makes testing that module easier (by setting a value higher than 1, so the initial cyard doesn't prevent the AI from using the module until it loses that cyard, at which point it may fail to finish a replacement due to factory destruction or lack of funds).

**Review by pchote — APPROVED** [review]

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/517fe4938d9ecb6fa33a5ba9fb72359cda870577
